hw4_1/hw4_1/Program.cs

```diff
@@ -0,0 +1,300 @@
+﻿namespace hw4_1
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            string path = "..\\..\\..\\..\\hw4_1.txt";
+            if (!File.Exists(path))
+            {
+                Console.WriteLine("Incorrect file path!");
+                return;
+            }
+
+            string expression;
+            using (StreamReader reader = new StreamReader(path))
+            {
+                if ((expression = await reader.ReadToEndAsync()) == null)
+                {
+                    Console.WriteLine("Empty file!");
+                    reader.Close();
+                    return;
+                }
+            }
+            
+            Console.WriteLine("Expression in file: " + expression);
+            try
+            {
+                ExpressionTree tree = new ExpressionTree(expression);
+            }
+            catch (Exception)
+            {
+                Console.WriteLine("File is incorrect!");
+                return;
+            }
+
+            ExpressionTree newCorrectTree = new ExpressionTree(expression);
+            Console.Write("Expression in expression tree: ");
+            newCorrectTree.PrintExpressionInTree();
+            try
+            {
+                newCorrectTree.CalculateExpression();
+            }
+            catch (DivideByZeroException e)
+            {
+                Console.WriteLine("\nDivision by zero!");
+                Console.WriteLine("e.StackTrace:\n" + e.StackTrace);
+                return;
+            }
+
+            float result = newCorrectTree.CalculateExpression();
+            Console.WriteLine("\nResult: " + result);
+            Console.WriteLine();
+        }
+    }
+
+    public class ExpressionTree
+    {
+        private OperandVertex Root;
+
+        public bool IsOperator(string symbol)
+        {
+            return symbol == "+" || symbol == "-" || symbol == "*" || symbol == "/";
+        }
+
+        public class OperationVertex
+        {
+            public string Operation;
+            public OperandVertex LeftOperand;
+            public OperandVertex RightOperand;

```

Comment:
```suggestion
            public IVertex LeftOperand;
            public IVertex RightOperand;
```
Неудачная иерархия классов.
Предполагалось, что получится интерфейс или абстрактный класс, от которого наследуются операции и операнды. Тогда в операции можно хранить операнды единообразно, вне зависимости от того, это числа или другие операции.